examples/seismic/acoustic/operators.py

```diff
@@ -1,10 +1,44 @@
+from sympy import sign
+
 from devito import Eq, Operator, Function, TimeFunction, Inc, solve
+from devito.symbolics import retrieve_functions, INT
 from examples.seismic import PointSource, Receiver
 
 
-def laplacian(field, m, s, kernel):
+def freesurface(model, eq):
+    """
+    Generate the stencil that mirrors the field as a free surface modeling for
+    the acoustic wave equation.
+
+    Parameters
+    ----------
+    model : Model
+        Physical model.
+    eq : Eq
+        Time-stepping stencil (time update) to mirror at the freesurface.
     """
-    Spacial discretization for the isotropic acoustic wave equation. For a 4th
+    lhs = eq.lhs
+    rhs = eq.rhs.evaluate
+    # Get vertical dimension and corresponding subdimension
+    z = model.grid.dimensions[-1]
+    zfs = model.grid.subdomains['fsdomain'].dimensions[-1]
+
+    # Functions present in the stencil
+    funcs = retrieve_functions(rhs)
+    mapper = {}
+    # Mirror negative indices
+    # TODO: Make a proper "mirror_indices" tool function
+    for f in funcs:
```

Comment:
I see the `TODO` - but if I'm reading this correctly we're messing with values in the halo to implement this atm? Is that something we really want to do? (Please correct me if I'm misunderstanding).

Reply:
Oh, or is the sponge still there and values are being manipulated in that?

Reply:
This is not touching the halo at all no. This is just turning any indices smaller than symbolic min into its mirror i.e u[x-1] => u[ abs(x-1)]. The halo is actually unnecessary then since nothing is ever accessed in it.

Reply:
Oh, ok, I was initially thrown by the `if ... < 0`. So you're creating a 'boundary layer' of length `so` to mirror the values in?

Reply:
Yes, the top `so` gridpoints are the layer in which indices are mirrored if negative. Ultimately (after some work) this would be basically autowrapping at the edge of any domain which would remove the need of a halo.

Reply:
And the boundary layer could also be completely removed by modifying the fd stencil appropriately -- easily doable for iso-acoustic, but probably too much pain to do nicely once this is generalized.

Reply:
>And the boundary layer could also be completely removed by modifying the fd stencil appropriately

That'd be really ideal yes but not really easy on the symbolic side as now you would have an different stencil (coeffs and radius) per distance to the boundary so wouldn't really be doable with a subdomain that expect the same equation for all points in the domain. Even for iso-acoustic, it's fairly trivial to do it by hand but not as a generic symbolic rule (unless I missed something).

Reply:
> Even for iso-acoustic, it's fairly trivial to do it by hand but not as a generic symbolic rule (unless I missed something).

For flat boundaries we can make use of custom FD coefficients to do it pretty easily.